src/missions/use-after-free-control-flow/temporal-mission.c

```diff
@@ -0,0 +1,280 @@
+ */
+#include <assert.h>
+#include <ctype.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sys/queue.h>
+
+#if defined(CAPREVOKE) && __CHERI_PURE_CAPABILITY__
+#include <sys/caprevoke.h>
+static void __attribute__((used)) *check_caprevoke = caprevoke;
+extern void malloc_revoke(void);
+static void __attribute__((used)) *check_malloc_revoke = malloc_revoke;
+#endif
+
+struct farm;
+struct crop;
+
+static void
+success(struct farm *f)
+{
+	fprintf(stderr, "Exploit successful: computer bought the farm!\n");
+	exit(0);
+}
+
+TAILQ_HEAD(cropq, crop);
+
+struct farm {
+	struct cropq	cropq;
+	struct crop	*cursor;
+};
+
+struct crop {
+	void	(* volatile describe)(struct crop *);
+	TAILQ_ENTRY(crop)	farm_cropq;
+};
+
+static_assert(sizeof(struct crop) == sizeof(struct farm),
+  "Structure size mismatch");
+
+union ufo {
+	void *ptrA;
+	char buf[sizeof(struct crop)];
+};
+
+static_assert(sizeof(struct crop) == sizeof(union ufo),
+  "Structure size mismatch");
+
+static void
+descr_1(struct crop *c)
+{
+	fprintf(stderr, "Chelan at %p\n", c); /* Pacific Northwest */
+}
+
+static void
+descr_2(struct crop *c)
+{
+	fprintf(stderr, "Colt at %p\n", c); /* United Kingdom */
+}
+
+
+static unsigned int cid;
+static union ufo *ufo; /* No state other than what's been abducted */
+
+static const size_t nfarm = 4;
+struct farm *farmp[nfarm];
+
+static void
+rm_farm(int fix)
+{
+	struct farm *f = farmp[fix];
+
+	farmp[fix] = NULL;
+	if (f != NULL) {
+		struct crop *c, *tc;
+
+		TAILQ_FOREACH_SAFE(c, &f->cropq, farm_cropq, tc) {
+			TAILQ_REMOVE(&f->cropq, c, farm_cropq);
+			free(c);
+		}
+
+		fprintf(stderr, "Tear down farm (index %d) at %p\n", fix, f);
+		free(f);
+	}
+}
+
+static struct farm *
+mk_farm(int fix)
+{
+	struct farm *f;
+
+	rm_farm(fix);
+
+	f = malloc(sizeof(struct farm));
+	assert(f != NULL); /* Surely infinite memory */
+
+	TAILQ_INIT(&f->cropq);
+	f->cursor = NULL;
+
+	fprintf(stderr, "New farm (index %d) at %p\n", fix, f);
+	return f;
+}
+
+static void
+rm_crop(struct farm *f, struct crop *c)
+{
+	fprintf(stderr, "Del crop at %p\n", c);
+	TAILQ_REMOVE(&f->cropq, c, farm_cropq);
+	free(c);
+}
+
+static struct crop *
+mk_crop(struct farm *f)
+{
+	struct crop *c;
+
+	c = malloc(sizeof(struct crop));
+	assert(c != NULL);
+
+	if (f->cursor != NULL) {
+		/* Inherit description of current cursor */
+		c->describe = f->cursor->describe;
+	} else if (cid & 1)
+		c->describe = descr_1;
+	else
+		c->describe = descr_2;
+	cid++;
+
+	TAILQ_INSERT_HEAD(&f->cropq, c, farm_cropq);
+
+	fprintf(stderr, "New crop at %p\n", c);
+
+	return c;
+}
+
+static void
+rm_ufo(void)
+{
+	if (ufo != NULL) {
+		fprintf(stderr, "Del UFO at %p\n", ufo);
+		free(ufo);
+	}
+}
+
+int
+main(void)
+{
+	int c;
+	size_t fix = 0;
+
+	/* Program startup in GDB is not the most celeritous thing */
+#if defined(CAPREVOKE) && __CHERI_PURE_CAPABILITY__
+	fprintf(stderr, "Ready (CHERI-RISC-V & heap safety)\n");
+#elif __CHERI_PURE_CAPABILITY__
```

Comment:
Missing `defined(...)`

Reply:
Yes, and, while thinking about preprocessor defines, I've made caprevoke the default and flipped to a `-DNO_CAPREVOKE` if you don't want it.

Reply:
Indeed.  While here, I've just dropped `defined(CAPREVOKE)` everywhere.